[WIP] Deprecate properties in favor of features

andy-sweet opened the pull request:
# Fixes / Closes

Closes #3911

# Description

This is a joint effort from me and @kevinyamauchi to deprecate the public API around `properties` state/parameters across all layers and related utilities, in favor of `features`. We also had to update some implementation details to avoid triggering the deprecation warnings in normal usage of napari.

Many of the code changes are in tests. These are mostly updated to maintain behaviors, but use `features` instead of `properties`. We also ensured there are some tests to continue to check that `properties` are working as expected and also issue deprecation warnings.

# References

Part of #2866

andy-sweet committed:
Deprecate properties parameters and properties

andy-sweet committed:
Update examples to use features

andy-sweet committed:
Define some new tests

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet committed:
Filter deprecation warnings to pass tests

andy-sweet committed:
Remove any non-points deprecation

andy-sweet committed:
Revert more non-points files.

andy-sweet committed:
Clean up

andy-sweet committed:
Update examples

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet committed:
Fix example

andy-sweet committed:
Refactor current properties setter

andy-sweet committed:
Fix tests

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet committed:
Remove unnecessary changes

andy-sweet committed:
Fix broken tests

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet committed:
Update deprecation version number

andy-sweet committed:
[skip ci]

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet commented:
@kevinyamauchi : I picked this up and this is a little easier after merging https://github.com/napari/napari/pull/5646

Will mark this as ready for review when I think this is in a good place.

andy-sweet committed:
Progress with all tests passing

andy-sweet committed:
Explicit deprecation catches for all

andy-sweet committed:
Fix some tests

andy-sweet committed:
More specific deprecation

andy-sweet committed:
Fix warning emitters

andy-sweet commented:
Made some progress here. As predicted, most changes are to tests where I mostly updated existing tests to uses features instead of properties. I kept a few properties specific tests and verified that deprecation warnings are issued there.

I think the only failing tests are related to `properties` and `property_choices` being returned in the `Layer._get_state` dictionary, but also being used for initialization, so they trigger the parameter deprecation warning. While we could just capture/ignore those warnings, they represent a real issue for plugin developers and maybe others, which is that we have no way to indicate that those keys are deprecated in the dictionary. So I think now is the time to consider a solution for that general problem - the simplest I can think of is to extend `dict` and make it issue the warning when accessing those keys.

kevinyamauchi commented:
Thanks for the update @andy-sweet !

> I think the only failing tests are related to properties and property_choices being returned in the Layer._get_state dictionary, but also being used for initialization, so they trigger the parameter deprecation warning. While we could just capture/ignore those warnings, they represent a real issue for plugin developers and maybe others, which is that we have no way to indicate that those keys are deprecated in the dictionary. So I think now is the time to consider a solution for that general problem - the simplest I can think of is to extend dict and make it issue the warning when accessing those keys.

I agree we should avoid just suppressing the warning because of the impact on plugin developers you mentioned. Extending `dict` seems reasonable to me.

andy-sweet committed:
Add special dict for layer state

andy-sweet commented on napari/layers/utils/_state_dict.py:
```diff
@@ -0,0 +1,19 @@
+from typing import Any
+
+
+class LayerStateDict(dict[str, Any]):
```
Any expertise on the easiest/safest way to subclass `dict`? Seems like `UserDict` used to be the recommended way, but that might not be the case anymore?

andy-sweet committed:
Simplify typing

andy-sweet commented on napari/layers/utils/_state_dict.py:
```diff
@@ -0,0 +1,19 @@
+from typing import Any
+
+
+class LayerStateDict(dict):
```
Any expertise on the easiest/safest way to subclass `dict`? Ideally with specialized typing. Seems like `UserDict` used to be the recommended way, but that might not be the case anymore?

andy-sweet commented:
> Extending `dict` seems reasonable to me.

I made a rough attempt at doing that. This is still a WIP, but if anyone had thoughts, I'd be receptive.

andy-sweet committed:
Refactor deprecation messages

andy-sweet committed:
Fix builtin points reader

kevinyamauchi commented:
Quick question: should the line below be updated to use `self._features.properties()`?
https://github.com/napari/napari/blob/8a1ad681205e577d648e45f884f2720873880064/napari/layers/points/points.py#L1286-L1287

```python
        self._edge._refresh_colors(self._feature_table.properties(), update_color_mapping)
        self._face._refresh_colors(self._feature_table.properties(), update_color_mapping)
```

andy-sweet commented:
> Quick question: should the line below be updated to use `self._features.properties()`?
> 
> https://github.com/napari/napari/blob/8a1ad681205e577d648e45f884f2720873880064/napari/layers/points/points.py#L1286-L1287
> 
> ```python
>         self._edge._refresh_colors(self._feature_table.properties(), update_color_mapping)
>         self._face._refresh_colors(self._feature_table.properties(), update_color_mapping)
> ```

Definitely. I guess this line isn't covered by tests.

kevinyamauchi commented:
> Definitely. I guess this line isn't covered by tests?

😬 guess not

kevinyamauchi committed:
[WIP] deprecate properties in vectors layer (#58)

# Description
This PR deprecates properties in the vectors layer based on the approach
in https://github.com/napari/napari/pull/5477

# References
Should be merged into https://github.com/napari/napari/pull/5477

andy-sweet committed:
Deprecate properties in tracks

andy-sweet committed:
Merge branch 'deprecate-properties-tracks' into deprecate-properties

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet committed:
Deprecate properties in labels

andy-sweet committed:
Fix vector and other tests

andy-sweet commented on napari/_qt/layer_controls/qt_vectors_controls.py:
```diff
@@ -55,15 +55,14 @@ class QtVectorsControls(QtLayerControls):
             self.change_edge_color_property
         )
-        self.color_prop_box.addItems(color_properties)
+        self.color_prop_box.addItems(self.layer.features.columns)
```
Considered renaming the names of these attributes from `prop` to `feature`, but I think this technically part of the napari package public API. I.e. I think someone can access it with something like:

```python
vectors = viewer.add_vectors()
qt_viewer = QtViewer(viewer)
qt_viewer.controls.widgets[vectors].color_prop_box
```

though I'm not sure if that's common or expected.

kevinyamauchi replied:
hmm good point. I'm not sure the best thing to do here. It feels okay to leave as-is, but I'm not sure.

andy-sweet replied:
I'm going to leave them alone for now. I don't think the variable names here are critical and I don't want to break anyone unnecessarily.

kevinyamauchi committed:
deprecate properties in Shapes (#59)

# Description
This PR deprecates properties in Shapes.

kevinyamauchi commented:
Hey @andy-sweet ! I made a small proposal for adding a note about the deprection in the points layer docstrings here: https://github.com/andy-sweet/napari/pull/60

kevinyamauchi committed:
add points deprecation docstring (#60)

# Description
This PR contains a proposal for adding deprecation notes to the
docstrings in the Points layer.

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet committed:
Merge branch 'deprecate-properties' of github.com:andy-sweet/napari into deprecate-properties

andy-sweet committed:
Add deprecations to docstrings

andy-sweet commented:
@kevinyamauchi : I changed the docstring you suggested a little to use the `.. deprecated:: 0.5.0 ...` form from [the numpydoc guide](https://numpydoc.readthedocs.io/en/latest/format.html#deprecation-warning).

For example:

```python
"""
...
        .. deprecated:: 0.5.0
            properties was deprecated in version 0.5.0 and will be removed in 0.6.
            Please use features instead.
...
"""

I used that form for all the attributes, parameters, and property (getters) that are currently documented. What do you think?

andy-sweet commented on napari/layers/shapes/shapes.py:
```diff
@@ -545,25 +572,27 @@ def __init__(
-                elem_name="face_color",
-                default="black",
-            )
+        # if len(data) > 0:
```
@kevinyamauchi : did you comment this out? If so, do you remember why? I don't remember doing it, but can check! Can also uncomment and see what breaks.

andy-sweet replied:
I uncommented this and fixed it by using `len(self.features.columns) == 0` instead of `self.features.empty`. Note that `DataFrame.empty` is `True` if there are either 0 rows or 0 columns.

I also had to remove the two lines you added calling `_initialize_current_color` and its definition.

kevinyamauchi replied:
Ah sorry, that was me. 😬  Thanks for the fix!

andy-sweet committed:
Some simple cleanup

andy-sweet committed:
Fix shape current colors

andy-sweet committed:
Remove unused function

andy-sweet committed:
Check for more deprecations

andy-sweet committed:
Handle deprecated state in labels test

andy-sweet committed:
Use shape instead of length of columns

andy-sweet committed:
Merge branch 'main' into deprecate-properties

kevinyamauchi commented:
> I used that form for all the attributes, parameters, and property (getters) that are currently documented. What do you think?

I like it! I didn't realize that was a thing. Good catch.

andy-sweet committed:
Merge branch 'main' into deprecate-properties

andy-sweet committed:
Add deprecating dict

andy-sweet committed:
Use deprecating dict for points deprecation

andy-sweet committed:
More clean up

andy-sweet committed:
Add explanatory comment

andy-sweet committed:
Fix test

andy-sweet committed:
Remove typing on layer data tuple method

andy-sweet committed:
Fix some tests

andy-sweet committed:
Temporarily ignore failed test

andy-sweet committed:
Store deprecated values separately

andy-sweet committed:
Add set, del, and contains

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Fix typing with cast

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Use FutureWarning instead of DeprecationWarning

andy-sweet committed:
Fix test

andy-sweet committed:
Use pytest warns instead of raises

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Merge branch 'feat-layer-state-deprecation' into deprecate-properties

andy-sweet committed:
Pass tests

andy-sweet committed:
Remove state dict

andy-sweet committed:
Format

andy-sweet commented:
I brought this one back from the dead and based it on https://github.com/napari/napari/pull/6976

andy-sweet committed:
Update signature of get_state on all layers

andy-sweet committed:
Update signature of get_state on all layers

andy-sweet committed:
Make DeprecatingDict public

andy-sweet committed:
Move and add tests

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Merge branch 'feat-layer-state-deprecation' into deprecate-properties

andy-sweet committed:
Use FutureWarning instead of DeprecationWarning

Czaki committed:
Add more general mechanism for deprecating properies in state (#69)

# Description

Provide more generic mechanism to deprecate states, to have a single
source of truth

---------

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

andy-sweet committed:
Refactor deprecated properties

Czaki committed:
Merge branch 'main' into feat-layer-state-deprecation

jni committed:
Fix importerror and message for lxml_html_clean (#7017)

Closes #6798

See https://github.com/napari/napari/issues/6798#issuecomment-2181495221
for discussion.

psobolewskiPhD committed:
[UX/UI] flip the default for the reader plugin dialog to *not* have the Remember box checked (#7016)

# References and relevant issues
Closes: https://github.com/napari/napari/issues/7010

# Description
- flip the defaults for the reader plugin dialog for the `Remember this
plugin` checkbox from True (checked) to False (unchecked)
- flip the tests to match

psobolewskiPhD committed:
Make Shift-Up/Down in layer list expand and contract selection (#6606)

# References and relevant issues
Closes https://github.com/napari/napari/issues/6604

# Description
Qt behavior is for Shift-Up/Down to only expand selection. This isn't
consistent with file managers like macOS Finder or most (all?) text
editors, when selecting lines. E.g. Shift-Up adds next item to
selection, but then Shift-Down will remove it--and vice versa.
As a result in this PR I first prevent Qt from handling Shift-Up and
Shift-Down.
Then I tweak the (unused) `shift=True` behavior of layers.select_next
(and thus select_previous)
Finally I add Viewer keybindings to handle Shift-Up/Down to
select_next(shift=True) and select_previous(shift=True)
Note I hard-code these, so can't be user modified, like some of our
other base keybinds.

New behavior:


https://github.com/napari/napari/assets/76622105/aec778f7-42bc-4dd7-adf5-04780232f0ee

dalthviz committed:
Add tests for viewer scale bar attributes (`colored` and `ticks`) (#7006)

# References and relevant issues

Closes #6990 

# Description

Add test to check visibility of scale bar when changing viewer `colored`
and `ticks` attribute

Czaki committed:
Again fix translate of overlays (this time with tests) (#6927)

# References and relevant issues
closes #6897

fix again bug from #6633

# Description

This PR is fixing again problems with transforms of overlays, this time
it contains basic tests.

---------

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>
Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

Czaki committed:
Deprecated properies that handles also Image class (#70)

# References and relevant issues

https://github.com/napari/napari/pull/6976

# Description

I just realized that state is a set of arguments to the constructor, so
we should not depend on `add_deprecated_property` but on the results of
`rename_argument`

The `rename_argument` already defines `class.__init_._rename_argument`
attribute that is used for generation of `add_xxx` methods.

This approach of changes also handle renaming of `interpolation` to
`interpolation2d` in the `Image` class.

As an improvement, PR may contain an update of `rename_argument` to use
`DeprecatedProperty.message` in `warnings.warn` used in
`rename_argument`

andy-sweet committed:
Address PR feedback

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Revert unneeded changes

andy-sweet committed:
Rename property to parameter

jni committed:
add match for deprecation message testing

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

andy-sweet committed:
Use separate pytest.warns

andy-sweet committed:
Address PR feedback and simplify

andy-sweet committed:
Merge branch 'feat-layer-state-deprecation' of github.com:andy-sweet/napari into feat-layer-state-deprecation

andy-sweet committed:
Use match in other tests

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Bring back named tuple for renamed attribute

andy-sweet committed:
Only handle renames

andy-sweet committed:
Refactor implementation

andy-sweet committed:
Apply suggestions from code review

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

andy-sweet committed:
Merge branch 'feat-layer-state-deprecation' of github.com:andy-sweet/napari into feat-layer-state-deprecation

andy-sweet committed:
Update docstring return types

andy-sweet committed:
Privatize deprecating dict

andy-sweet committed:
Implement with UserDict to get pop, get, etc.

andy-sweet committed:
Only rename key if in renamed

andy-sweet committed:
Merge branch 'feat-layer-state-deprecation' into deprecate-properties-with-new-dd

andy-sweet committed:
Fix names of tests

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Merge branch 'feat-layer-state-deprecation' into deprecate-properties-with-new-dd

andy-sweet committed:
Raise when mutating deprecated

andy-sweet committed:
Many fixes

andy-sweet committed:
Fix built-in points export

andy-sweet committed:
Fix typing

andy-sweet committed:
Fix shapes warnings for numpy 2

andy-sweet committed:
Fix docs

andy-sweet committed:
Add test for general deprecations

andy-sweet committed:
Merge branch 'main' into deprecate-properties

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

andy-sweet commented:
I'm planning to split this XXL PR into two or three parts to help review the changes.
